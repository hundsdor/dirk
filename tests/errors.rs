use std::{path::PathBuf, process::Command};

use insta::{assert_snapshot, with_settings, Settings};
use tempdir::TempDir;
use test_case::test_case;

#[test_case("blueprint")]
#[test_case("missing_provides")]
#[test_case("use_inject_on_impl")]
#[test_case("component_on_impl")]
#[test_case("component_type_mismatch")]
#[test_case("component_type_mismatch_generics")]
#[test_case("component_missing_dependency")]
#[test_case("component_missing_binding")]
#[test_case("provides_on_trait")]
#[test_case("provides_on_empty_impl")]
#[test_case("provides_on_impl_with_more_than_one_function")]
#[test_case("provides_invalid_return_type")]
fn test_errors(name: &str) {
    let mut cmd = Command::new(env!("CARGO"));

    cmd.arg("run");
    cmd.arg("--example");
    cmd.arg(name);

    cmd.env("RUSTFLAGS", "--cap-lints=allow");

    let mut current_dir = PathBuf::new();
    current_dir.push("test-data");
    current_dir.push("errors");

    cmd.current_dir(&current_dir);

    let target_dir = TempDir::new_in(env!("CARGO_TARGET_TMPDIR"), name).unwrap();
    cmd.env("TARGET_DIR", target_dir.path());

    let output = cmd
        .output()
        .unwrap_or_else(|_| panic!("Failed to execute test {name}"));

    let mut settings = Settings::clone_current();

    let filter_block_package_cache = r".*Blocking waiting for file lock on package cache.*\n";
    settings.add_filter(filter_block_package_cache, "");

    let filter_block_build_directory = r".*Blocking waiting for file lock on build directory.*\n";
    settings.add_filter(filter_block_build_directory, "");

    let filter_finished_targets = r".*Finished .* target\(s\) in .*s";
    settings.add_filter(filter_finished_targets, "Finished compiling target(s)");

    let filter_compiling = r".*Compiling .*\n";
    settings.add_filter(filter_compiling, "");

    let filter_updating_index = r".*Updating crates.io index\n";
    settings.add_filter(filter_updating_index, "");

    let pretty = format!(
        "Status {:?}\n\nStdout:\n{}\n\nStderr:\n{}",
        output.status,
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr)
    );

    settings.bind(|| {
        with_settings!({snapshot_suffix => name}, {
            assert_snapshot!(pretty);
        });
    });
}
