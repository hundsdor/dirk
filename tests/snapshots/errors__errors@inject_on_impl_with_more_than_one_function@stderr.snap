---
source: tests/errors.rs
expression: "String::from_utf8_lossy(&output.stderr)"
---
error: #[*_inject] is supposed to be placed on an impl block containing one single function - found 2 functions instead
  --> examples/inject_on_impl_with_more_than_one_function.rs:53:1
   |
53 | / impl<H: Heater, P: Pump> CoffeeMaker<H, P> {
54 | |     fn new(
55 | |         logger: Arc<RwLock<CoffeeLogger>>,
56 | |         heater: Rc<RefCell<H>>,
...  |
68 | |     }
69 | | }
   | |_^

error[E0433]: failed to resolve: use of undeclared type `StaticFactoryCoffeeMaker`
  --> examples/inject_on_impl_with_more_than_one_function.rs:36:28
   |
36 |         maker: static_bind(CoffeeMaker<ElectricHeater, ThermoSiphon<ElectricHeater>>) [logger, heater, pump]
   |                            ^^^^^^^^^^^ use of undeclared type `StaticFactoryCoffeeMaker`

For more information about this error, try `rustc --explain E0433`.
error: could not compile `errors` (example "inject_on_impl_with_more_than_one_function") due to 2 previous errors
